### PR TITLE
PIRE bindings in Cython

### PR DESCRIPTION
**Pull request by moskupols**
Reimplementation of #4 in Cython

### CODE REVIEW

File: bindings/python/stub.pxd [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+
+    cdef cppclass auto_ptr[T]:
+        auto_ptr()
+
+        T* get()
+        void reset(T*)
+
+
+cdef extern from "pire/pire.h" namespace "Pire":
+    ctypedef vector yvector
```
Comment (akindyakov) [inline bindings/python/stub.pxd:26]: А зачем ты вектор и строку переименовываешь в y\* ?

File: bindings/python/render.py [hunk 2]
```diff
@@ -31,10 +31,10 @@
         "Minimize",
     ],
     "scanners": [
-        "Scanner",
-        "NonrelocScanner",
-        "ScannerNoMask",
-        "NonrelocScannerNoMask",
+        ("Scanner", "size_t"),
+        ("NonrelocScanner", "size_t"),
```
Comment (akindyakov) [inline bindings/python/render.py]: Может ли тут понадобиться что то отличное от size_t?

File: bindings/python/tests/run-bench.in [hunk 3]
```diff
@@ -0,0 +1,3 @@
+#!@SHELL@
+
+exec "@builddir@"/run-with-pypath @PYTHON@ "@srcdir@"/bench.py -f "@top_srcdir@"/tools/bench/test_file -m 5000 -p ".*[net]$"
```
Comment (akindyakov) [inline bindings/python/tests/run-bench.in]: Может через \ в столбик?

File: bindings/python/tests/bench.py [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+        for line in self.lines:
+            self.caller.Call(line)
+
+
+def make_argparser():
+    parser = argparse.ArgumentParser()
+    parser.add_argument(
+        "-f", "--file",
+        type=argparse.FileType("r"),
+        default=sys.stdin,
```
Comment (akindyakov) [inline bindings/python/tests/bench.py:63]: help бы опциям, м?

File: bindings/python/tests/bench.py [hunk 5]
```diff
@@ -0,0 +1,84 @@
...
+    options = make_argparser().parse_args()
+
+    lines = map(str.strip, options.file.readlines())
+    lines *= options.multiply_input
+
+    for caller_name, caller_class in CALLERS.items():
+        global runner
+        runner = TestRunner(caller_class(options.pattern), lines)
+        print caller_name
+        cProfile.run("runner.Run()", sort="cumtime")
```
Comment (akindyakov) [inline bindings/python/tests/bench.py:94]: А где ты принтишь время работы?

File: bindings/python/stub.pxd (hunk 1 above)
Comment (moskupols) [inline bindings/python/stub.pxd:26]: Чтобы вместо std::vector использовать Pire::yvector, который там переопределён, чтобы было проще подменять реализацию

File: bindings/python/render.py (hunk 2 above)
Comment (moskupols) [inline bindings/python/render.py]: Да. У `CapturingScanner` состояние -- это класс, который умеет говорить позицию захваченной группы. В более позднем коммите я перенёс `"size_t"` в аргумент по умолчанию у `ScannerSpec`.

File: bindings/python/tests/bench.py (hunk 5 above)
Comment (moskupols) [inline bindings/python/tests/bench.py:94]: Его cProfile.run и принтит

File: bindings/python/tests/run-bench.in (hunk 3 above)
Comment (moskupols) [inline bindings/python/tests/run-bench.in]: Поправил

File: bindings/python/tests/bench.py (hunk 4 above)
Comment (moskupols) [inline bindings/python/tests/bench.py:63]: Добавил

**Comment by moskupols** [comment]
Вроде бы, покрыл всё, кроме функционала PIRE_INLINE, но его можно сделать на месте, например, тем же mako
UPD: на самом деле, я вспомнил, пока не покрыты загрузка сканеров через mmap и  низкоуровневые операции с Fsm.

**Comment by starius** [comment]
Please make sure it builds correctly with `make -j 8`.
Missing dependencies between build artefacts can cause parallel build to fail.

File: bindings/python/pire.pyx.mako [hunk 6]
```diff
@@ -0,0 +1,369 @@
...
+        self.scanner_impl.Save(&stream)
+        return stream.GetStr()
+
+    Load = _${Scanner}_Load
+
+    def InitState(self):
+        return ${Scanner}State.__new__(${Scanner}State, self)
+
+    % if "Glue" not in spec.ignored_methods:
+    def GluedWith(self, ${Scanner} rhs not None, size_t max_size=0):
```
Comment (starius) [inline bindings/python/pire.pyx.mako:257]: What is the purpose of `not None`?

File: bindings/python/pire.pyx.mako [hunk 7]
```diff
@@ -0,0 +1,369 @@
...
+    % for method in ["LongestSuffix", "ShortestSuffix"]:
+    def ${method}(self, bytes line not None):
+        cdef:
+            const char* rbegin = end(line) - 1
+            const char* rend = begin(line) - 1
+            const char* suffix_begin
+        suffix_begin = impl.${method}(self.scanner_impl, rbegin, rend)
+        if suffix_begin != NULL:
+            return suffix_begin - rend
+        return None
```
Comment (starius) [inline bindings/python/pire.pyx.mako]: IMHO,

``` python
if suffix_begin == NULL:
    return None
return suffix_begin - rend
```

To put `None` near `NULL`.

File: bindings/python/pire.pyx.mako [hunk 8]
```diff
@@ -0,0 +1,369 @@
...
+    cdef inline impl.yauto_ptr[impl.Options] Convert(self):
+        return impl.ConvertFlagSetToOptions(<impl.FlagSet>self.option_set)
+
+    cdef inline void Apply(self, Lexer lexer):
+        cdef impl.yauto_ptr[impl.Options] converted = self.Convert()
+        converted.get().Apply(cython.operator.dereference(lexer.lexer_impl))
+
+
+% for option in OPTIONS:
+${option} = Options.__new__(Options, {impl.${option}})
```
Comment (starius) [inline bindings/python/pire.pyx.mako]: What `{` and `}` mean in `{impl.${option}}`? If they create a set, replace with `set([...])`, please.

File: bindings/python/render.py [hunk 9]
```diff
@@ -0,0 +1,113 @@
...
+        "BeginMark",
+        "EndMark",
+        "MaxCharUnaligned",
+        "MaxChar",
+    ],
+}
+
+
+def make_argparser():
+    parser = argparse.ArgumentParser()
```
Comment (starius) [inline bindings/python/render.py]: IMHO, add docstring to the file and

``` python
parser = argparse.ArgumentParser(
    description=__doc__,
    formatter_class=argparse.ArgumentDefaultsHelpFormatter,
)
```

File: bindings/python/tests/bench.py [hunk 10]
```diff
@@ -0,0 +1,91 @@
...
+        self.caller = caller
+        self.lines = lines
+
+    def Run(self):
+        for line in self.lines:
+            self.caller.Call(line)
+
+
+def make_argparser():
+    parser = argparse.ArgumentParser()
```
Comment (starius) [inline bindings/python/tests/bench.py]: IMHO, add docstring to the file and

``` python
parser = argparse.ArgumentParser(
    description=__doc__,
    formatter_class=argparse.ArgumentDefaultsHelpFormatter,
)
```

File: bindings/python/tests/pire_test.py [hunk 11]
```diff
@@ -0,0 +1,401 @@
...
+        assert scanner_class == type(pire.Fsm().Compile(scanner_class))
+
+    def test_fsm_is_copy_constructible(self):
+        fsm = pire.Fsm().Append("ab")
+        fsm_copy = pire.Fsm(fsm)
+        assert fsm_copy is not fsm
+        check_equivalence(
+            fsm.Compile(),
+            fsm_copy.Compile(),
+            ["", "a", "ab", "ab-", "-"]
```
Comment (starius) [inline bindings/python/tests/pire_test.py]: ,

File: bindings/python/tests/pire_test.py [hunk 12]
```diff
@@ -0,0 +1,401 @@
...
+        fsm.AppendSpecial(pire.BeginMark)
+        fsm.Append('a')
+        fsm.AppendSpecial(pire.EndMark)
+
+        scanner = scanner_class(fsm)
+        state = scanner.InitState()
+
+        check_state(state.Begin(), final=False, dead=False)
+        check_state(state.Run('a'), final=False, dead=False)
+        check_state(state.Step(pire.EndMark), final=True)
```
Comment (starius) [inline bindings/python/tests/pire_test.py:109]: What about `dead=...` for `state.Step(pire.EndMark)`?

File: bindings/python/tests/pire_test.py [hunk 13]
```diff
@@ -0,0 +1,401 @@
...
+        fsm.AppendStrings(["abc", "de"])
+        check_scanner(
+            scanner_class(fsm),
+            accepts=["-abc", "-de"],
+            rejects=["-", "abc", ""],
+        )
+
+    def test_fsm_raises_when_one_of_appended_strings_is_empty(self):
+        fsm = pire.Fsm()
+        pytest.raises(Exception, fsm.AppendStrings, [""])
```
Comment (starius) [inline bindings/python/tests/pire_test.py]: IMHO,

``` python
with pytest.raises(Exception):
    fsm.AppendStrings([""])
with pytest.raises(Exception):
    fsm.AppendStrings(["nonempty", ""])
```

Should we replace `Exception` with a subclass?

File: bindings/python/tests/pire_test.py [hunk 14]
```diff
@@ -0,0 +1,401 @@
...
+
+    def test_scanner_finds_prefixes_and_suffixes(self, scanner_class):
+        fsm = pire.Lexer("-->").Parse()
+        any_occurence = scanner_class(~pire.Fsm.MakeFalse() + fsm)
+        first_occurence = scanner_class(~fsm.Surrounded() + fsm)
+        reverse_occurence = scanner_class(fsm.Reverse())
+
+        text = "1234567890 --> middle --> end"
+        assert 14 == first_occurence.LongestPrefix(text)
+        assert 11 == reverse_occurence.LongestSuffix(text[:14])
```
Comment (starius) [inline bindings/python/tests/pire_test.py:269]: Style

```
assert expected == observed
```

is better, but pytest's error output suggests that pytest uses

```
assert observed == expected
```

File: bindings/python/tests/pire_test.py [hunk 15]
```diff
@@ -0,0 +1,401 @@
...
+
+    def test_regexp_matches(self):
+        re = pire.Regexp("(foo|bar)+", pire.I)
+        assert "prefix fOoBaR suffix" in re
+        assert "bla bla bla" not in re
+        assert re.Matches("barfoo")
+
+    def test_regexp_honors_utf8(self):
+        re = pire.Regexp("^.$", pire.I | pire.UTF8)
+        assert "\x41" in re
```
Comment (starius) [inline bindings/python/tests/pire_test.py]: IMHO, `s/\x41/A/`.

File: bindings/python/tests/pire_test.py [hunk 16]
```diff
@@ -0,0 +1,401 @@
...
+        assert "\x81" not in re
```
Comment (starius) [inline bindings/python/tests/pire_test.py]: IMHO, add comments:
- valid UTF-8
- invalid UTF-8

and add test for "Я"

File: bindings/python/pire.pyx.mako (hunk 6 above)
Comment (moskupols) [inline bindings/python/pire.pyx.mako:257]: Without `not None`, the generated code would check the received parameter for something equivalent to `rhs is None or isinstance(rhs, Scanner)`, and would not raise exception for `None` passed here. With `not None` the check is just `isinstance(rhs, Scanner)`.

The typecheck is necessary to early-bind methods and attributes of `rhs`.

File: bindings/python/tests/pire_test.py [hunk 17]
```diff
@@ -0,0 +1,401 @@
...
+            check_scanner(
+                scanner,
+                accepts=[exactly_regexp, almost_regexp.encode("utf8")],
+            )
+
+    def test_capturing_trivial(self):
+        """
+        This is the "Trivial" test from tests/capture_ut.cpp.
+        """
+        lexer = pire.Lexer("google_id\\s*=\\s*[\'\"]([a-z0-9]+)[\'\"]\\s*;")
```
Comment (starius) [inline bindings/python/tests/pire_test.py]: IMHO, use `r"..."` to simplify.

File: bindings/python/tests/pire_test.py [hunk 18]
```diff
@@ -0,0 +1,401 @@
...
+
+    def test_counting_scanner_is_default_constructible(self):
+        fsm = pire.Fsm()
+        pire.CountingScanner()
+
+    def test_counting_scanner_raises_when_constructed_without_second_fsm(self):
+        fsm = pire.Fsm()
+        pytest.raises(Exception, pire.CountingScanner, fsm)
+        pytest.raises(Exception, pire.CountingScanner, pattern=fsm)
+        pytest.raises(Exception, pire.CountingScanner, sep=fsm)
```
Comment (starius) [inline bindings/python/tests/pire_test.py]: `with pytest.raises(Exception):`

(See above.)

File: bindings/python/tests/pire_test.py (hunk 12 above)
Comment (moskupols) [inline bindings/python/tests/pire_test.py:109]: `final=True` implies `dead=False`, and this logic is applied by `check_state` if `dead` argument is not specified or `None`, as well as similar logic for `final=False` implying `accepted_regexps=()`.

File: configure.ac [hunk 19]
```diff
@@ -97,13 +119,23 @@ AC_CONFIG_FILES([
...
 	tools/Makefile
 	tools/bench/Makefile
 	samples/Makefile
 	samples/inline/Makefile
 	samples/blacklist/Makefile
 	samples/pigrep/Makefile
+], [
+	chmod +x bindings/python/tests/run-with-pypath
+	chmod +x bindings/python/tests/run-bench
+	chmod +x bindings/python/tests/run-pytest
```
Comment (starius) [inline configure.ac:138]: Can you avoid this by `chmod +x`ing `bindings/python/tests/*.in`?

File: bindings/python/pire.pyx.mako (hunk 7 above)
Comment (moskupols) [inline bindings/python/pire.pyx.mako]: Done.

File: configure.ac (hunk 19 above)
Comment (moskupols) [inline configure.ac:138]: No, access mode of `.in` files is not preserved by autotools.

File: bindings/python/pire.pyx.mako (hunk 8 above)
Comment (moskupols) [inline bindings/python/pire.pyx.mako]: Done.

File: bindings/python/render.py (hunk 9 above)
Comment (moskupols) [inline bindings/python/render.py]: I added the description, but the formatter you suggest prints the following:

```
  -o OUTPUT, --output OUTPUT
                        Path to place the rendered source at (default: None)
```

IMHO this `default: None` is misleading, as `output` is a required parameter.

File: bindings/python/tests/bench.py (hunk 10 above)
Comment (moskupols) [inline bindings/python/tests/bench.py]: Done.

File: bindings/python/tests/pire_test.py (hunk 11 above)
Comment (moskupols) [inline bindings/python/tests/pire_test.py]: Fixed

File: bindings/python/tests/pire_test.py (hunk 14 above)
Comment (moskupols) [inline bindings/python/tests/pire_test.py:269]: Could you provide an example for which pytest differentiates between expected and observed? In simple cases it doesn't.

```
>       assert Say5() == 6
E       assert 5 == 6
E        +  where 5 = Say5()
```

```
>       assert 6 == Say5()
E       assert 6 == 5
E        +  where 5 = Say5()
```

```
>       assert Say5() == Say6()
E       assert 5 == 6
E        +  where 5 = Say5()
E        +  and   6 = Say6()
```

File: bindings/python/tests/pire_test.py (hunk 17 above)
Comment (moskupols) [inline bindings/python/tests/pire_test.py]: Done.

File: bindings/python/tests/pire_test.py (hunk 16 above)
Comment (moskupols) [inline bindings/python/tests/pire_test.py]: Done.

File: bindings/python/tests/pire_test.py (hunk 15 above)
Comment (moskupols) [inline bindings/python/tests/pire_test.py]: I added a comment that this is `"A"`. Actually, the fact that `\x41 < \x80` is IMHO more important in this context than the fact that `\x41` is `"A"`.

File: bindings/python/tests/pire_test.py (hunk 13 above)
Comment (moskupols) [inline bindings/python/tests/pire_test.py]: Done.

File: bindings/python/tests/pire_test.py (hunk 15 above)
Comment (starius) [inline bindings/python/tests/pire_test.py]: What is really important is that "\x41" (or "A") is a valid UTF-8. From "A" it is clearer.

File: bindings/python/pire.pyx.mako [hunk 20]
```diff
@@ -224,9 +229,19 @@ copy_reg.pickle(${Scanner}, _reduce_${Scanner})
 cdef class ${Scanner}(BaseScanner):
     cdef impl.${Scanner} scanner_impl
 
+    % if Scanner != "CountingScanner":
     def __cinit__(self, Fsm fsm=None):
         if fsm is not None:
             self.scanner_impl = impl.${Scanner}(fsm.fsm_impl)
+    % else:
+    def __cinit__(self, Fsm pattern=None, Fsm sep=None):
+        if pattern is not None or sep is not None:
```
Comment (sergey-v-galtsev) [inline bindings/python/pire.pyx.mako]: ИМХО (изменил if+if на if+elif и сообщение об ошибке):

```
        if pattern is not None and sep is not None:
            self.scanner_impl = impl.CountingScanner(pattern.fsm_impl, sep.fsm_impl)
        elif pattern is None or sep is None:
            raise ValueError(
                "Expected both pattern and separator or neither, got only one of them"
            )
```

Comment (moskupols) [inline bindings/python/pire.pyx.mako]: Ok

**Pull request merged**